src/Microsoft.AspNet.Mvc.TagHelpers/InputTagHelper.cs

```diff
@@ -389,15 +390,14 @@ private string GetInputType(ModelExplorer modelExplorer, out string inputTypeHin
         // A variant of TemplateRenderer.GetViewNames(). Main change relates to bool? handling.
         private static IEnumerable<string> GetInputTypeHints(ModelExplorer modelExplorer)
         {
-            var inputTypeHints = new string[]
+            if (!string.IsNullOrEmpty(modelExplorer.Metadata.TemplateHint))
             {
-                modelExplorer.Metadata.TemplateHint,
-                modelExplorer.Metadata.DataTypeName,
-            };
+                yield return modelExplorer.Metadata.TemplateHint;
+            }
 
-            foreach (string inputTypeHint in inputTypeHints.Where(s => !string.IsNullOrEmpty(s)))
+            if (!string.IsNullOrEmpty(modelExplorer.Metadata.DataTypeName))
             {
-                yield return inputTypeHint;
+                yield return modelExplorer.Metadata.DataTypeName;
```

Comment:
That's not always `true` in part because these state machines aren't tiny. Will need to measure both approaches if this one shows up high when profiling a page with 10 or so `<input>` tag helpers.